File: android/src/main/kotlin/ai/voqal/voqal_flutter/VoqalFlutterPlugin.kt
package ai.voqal.voqal_flutter

import ai.voqal.sdk.VoqalConfiguration
import ai.voqal.sdk.VoqalDelegate
import ai.voqal.sdk.VoqalHome
import ai.voqal.sdk.VoqalSDK
import ai.voqal.sdk.VoqalTheme
import android.app.Activity
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import android.util.Log
import androidx.activity.ComponentActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Thin Flutter bridge over the native Voqal Android SDK.
 *
 * The SDK launches its own full-screen experience, so the bridge only forwards
 * configuration, pushed credentials, and the prewarm/present calls. The
 * MethodChannel is async but [VoqalDelegate] reads the token synchronously, so
 * Dart pushes credentials down (refreshing them on rotation) and the delegate
 * serves an in-memory cache. The token is never logged, persisted, or returned
 * across the channel.
 */
class VoqalFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null

    // Latest credentials pushed from Dart. In-memory only — never persisted.
    @Volatile private var cachedToken: String = ""
    @Volatile private var cachedMetadata: String? = null

    private val delegate = object : VoqalDelegate {
        override fun getToken(): String = cachedToken
        override fun getMetadata(): String? = cachedMetadata
        override fun onUploadResult(result: String) {
            // No-op for v1 (matches iOS). Wire a Dart listener before forwarding.
        }

        override fun onError(error: Throwable) {
            // Log the Throwable only — never the token or metadata.
            Log.e("VoqalFlutter", "Voqal error", error)
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "voqal_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "setup" -> {
                val config = call.arguments as? Map<*, *>
                    ?: return result.error("bad_args", "setup expects a configuration map", null)
                VoqalSDK.setup(configuration(config))
                result.success(null)
            }

            "setCredentials" -> {
                val token = call.argument<String>("token")
                    ?: return result.error("bad_args", "setCredentials expects a token", null)
                cachedToken = token
                cachedMetadata = call.argument<String>("metadata")
                result.success(null)
            }

            "prewarm" -> {
                VoqalSDK.prewarm(delegate)
                result.success(null)
            }

            "present" -> {
                val host = activity as? ComponentActivity
                    ?: return result.error(
                        "no_activity",
                        "present requires a foreground ComponentActivity (FlutterActivity)",
                        null,
                    )
                VoqalSDK.present(host, delegate)
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun configuration(config: Map<*, *>): VoqalConfiguration {
        // Note: `observability` is intentionally ignored — the Android SDK
        // exposes no observability surface yet (documented follow-up).
        return VoqalConfiguration(
            requestId = config["requestId"] as? String ?: "prod-flutter",
            apiKey = config["apiKey"] as? String,
            agentUrl = config["agentUrl"] as? String,
            theme = theme(config["theme"] as? Map<*, *>),
            home = home(config["home"] as? Map<*, *>),
            hapticsEnabled = config["hapticsEnabled"] as? Boolean ?: true,
            conversationTimeoutSeconds = (config["conversationTimeoutSeconds"] as? Number)?.toLong() ?: 7200L,
            presentationStyle = when (config["presentationStyle"] as? String) {
                "fullScreen" -> VoqalConfiguration.PresentationStyle.FULL_SCREEN
                else -> VoqalConfiguration.PresentationStyle.SHEET
            },
            headerTitle = config["headerTitle"] as? String,
            headerIcon = decodeIcon(config["headerIconPngBase64"] as? String),
        )
    }

    /** Decodes a base64 PNG (the header logo, sent from Dart) to a Bitmap. */
    private fun decodeIcon(base64: String?): Bitmap? {
        if (base64.isNullOrEmpty()) return null
        return try {
            val bytes = Base64.decode(base64, Base64.DEFAULT)
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        } catch (e: IllegalArgumentException) {
            Log.w("VoqalFlutter", "Ignoring malformed header icon", e)
            null
        }
    }

    private fun theme(map: Map<*, *>?): VoqalTheme {
        if (map == null) return VoqalTheme()
        return VoqalTheme(
            accent = map["accent"] as? String ?: "#2d5bff",
            accent2 = map["accent2"] as? String,
            appearance = appearance(map["appearance"] as? String),
            radius = (map["radius"] as? Number)?.toFloat() ?: 20f,
        )
    }

    private fun home(map: Map<*, *>?): VoqalHome {
        if (map == null) return VoqalHome()
        @Suppress("UNCHECKED_CAST")
        val ctas = (map["pinnedCTAs"] as? List<String>) ?: emptyList()
        return VoqalHome(
            userName = map["userName"] as? String,
            pinnedCTAs = ctas,
            showAgentGlance = map["showAgentGlance"] as? Boolean ?: true,
        )
    }

    private fun appearance(value: String?): VoqalTheme.Appearance = when (value) {
        "light" -> VoqalTheme.Appearance.LIGHT
        "dark" -> VoqalTheme.Appearance.DARK
        else -> VoqalTheme.Appearance.AUTO
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

File: ios/voqal_flutter/Sources/voqal_flutter/VoqalFlutterPlugin.swift
import Flutter
import UIKit
import VoqalSDK

/// Thin Flutter bridge over the native VoqalSDK.
///
/// The SDK presents its own full-screen experience, so this bridge only carries
/// configuration, pushed credentials, and the prewarm/present calls — no view
/// embedding. The MethodChannel is async but the SDK's delegate reads the token
/// synchronously, so Dart pushes credentials down (and refreshes them on
/// rotation) and the delegate serves a native cache. The token is never logged,
/// persisted, or returned across the channel.
public class VoqalFlutterPlugin: NSObject, FlutterPlugin {

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "voqal_flutter", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(VoqalFlutterPlugin(), channel: channel)
  }

  /// Latest credentials pushed from Dart. The SDK delegate may read these off
  /// the main thread, so access is lock-guarded for cross-thread visibility.
  /// In-memory only — never persisted.
  private static let cacheLock = NSLock()
  private static var cachedToken = ""
  private static var cachedMetadata: String?

  private static func cacheCredentials(token: String, metadata: String?) {
    cacheLock.lock()
    cachedToken = token
    cachedMetadata = metadata
    cacheLock.unlock()
  }

  private static func currentToken() -> String {
    cacheLock.lock()
    defer { cacheLock.unlock() }
    return cachedToken
  }

  private static func currentMetadata() -> String? {
    cacheLock.lock()
    defer { cacheLock.unlock() }
    return cachedMetadata
  }

  /// Retained delegate — prewarm's background work needs it alive.
  private static let coordinator = Coordinator()

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "setup":
      guard let config = call.arguments as? [String: Any] else {
        result(badArgs("setup expects a configuration map"))
        return
      }
      DispatchQueue.main.async {
        VoqalSDKManager.shared.setup(configuration: Self.configuration(from: config))
        result(nil)
      }

    case "setCredentials":
      guard let args = call.arguments as? [String: Any],
        let token = args["token"] as? String
      else {
        result(badArgs("setCredentials expects a token"))
        return
      }
      Self.cacheCredentials(token: token, metadata: args["metadata"] as? String)
      result(nil)

    case "prewarm":
      DispatchQueue.main.async {
        VoqalSDKManager.shared.prewarm(delegate: Self.coordinator)
        result(nil)
      }

    case "present":
      DispatchQueue.main.async {
        VoqalSDKManager.shared.presentChat(
          from: Self.coordinator.getViewController(),
          delegate: Self.coordinator,
          animated: true)
        result(nil)
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Builds a `FlutterError` that never embeds caller-supplied credentials.
  private func badArgs(_ message: String) -> FlutterError {
    FlutterError(code: "bad_args", message: message, details: nil)
  }

  // MARK: - Config mapping

  private static func configuration(from config: [String: Any]) -> VoqalSDKConfiguration {
    var configuration = VoqalSDKConfiguration(
      requestId: config["requestId"] as? String ?? "prod-flutter")

    if let key = config["apiKey"] as? String { configuration.apiKey = key }
    if let urlString = config["agentUrl"] as? String, let url = URL(string: urlString) {
      configuration.agentURL = url
    }
    if let haptics = config["hapticsEnabled"] as? Bool {
      configuration.hapticsEnabled = haptics
    }
    if let timeout = config["conversationTimeoutSeconds"] as? Double {
      configuration.conversationTimeout = timeout
    } else if let timeout = config["conversationTimeoutSeconds"] as? Int {
      configuration.conversationTimeout = Double(timeout)
    }
    configuration.theme = theme(from: config["theme"] as? [String: Any])
    configuration.home = home(from: config["home"] as? [String: Any])
    configuration.observability = observability(from: config["observability"] as? [String: Any])
    switch config["presentationStyle"] as? String {
    case "fullScreen": configuration.presentationStyle = .fullScreen
    default: configuration.presentationStyle = .sheet
    }
    if let title = config["headerTitle"] as? String {
      configuration.strings.chatHeaderTitle = title
    }
    if let base64 = config["headerIconPngBase64"] as? String,
      let data = Data(base64Encoded: base64),
      let image = UIImage(data: data)
    {
      configuration.icons.chatHeaderIcon = image
    }
    return configuration
  }

  private static func theme(from dictionary: [String: Any]?) -> VoqalTheme {
    guard let dictionary else { return VoqalTheme() }
    let appearance: VoqalTheme.Appearance
    switch dictionary["appearance"] as? String {
    case "light": appearance = .light
    case "dark": appearance = .dark
    default: appearance = .auto
    }
    let radius: CGFloat
    if let value = dictionary["radius"] as? Double { radius = CGFloat(value) } else { radius = 20 }
    // VoqalSDK 1.3.3 replaced `fontName: String?` with `font: UIFont?`. Keep the Flutter API the
    // same (a font-family name string) and map it to a UIFont here — the SDK uses its family.
    let font = (dictionary["fontName"] as? String).flatMap { UIFont(name: $0, size: 16) }
    return VoqalTheme(
      accent: dictionary["accent"] as? String ?? "#2d5bff",
      accent2: dictionary["accent2"] as? String,
      appearance: appearance,
      font: font,
      radius: radius)
  }

  private static func home(from dictionary: [String: Any]?) -> VoqalHome {
    guard let dictionary else { return VoqalHome() }
    return VoqalHome(
      userName: dictionary["userName"] as? String,
      pinnedCTAs: dictionary["pinnedCTAs"] as? [String] ?? [],
      showAgentGlance: dictionary["showAgentGlance"] as? Bool ?? true)
  }

  private static func observability(from dictionary: [String: Any]?) -> VoqalObservabilityOptions {
    var options = VoqalObservabilityOptions()
    guard let dictionary else { return options }
    if let dsn = dictionary["dsn"] as? String { options.dsn = dsn }
    if let enabled = dictionary["enabled"] as? Bool { options.enabled = enabled }
    if let scrub = dictionary["scrubPII"] as? Bool { options.scrubPII = scrub }
    if let rate = dictionary["tracesSampleRate"] as? Double { options.tracesSampleRate = rate }
    if let environment = dictionary["environment"] as? String { options.environment = environment }
    return options
  }

  // MARK: - Delegate

  /// Serves the pushed credentials synchronously and resolves the on-screen
  /// view controller. Logs errors only — never the token.
  final class Coordinator: NSObject, VocalButtonDelegate {
    func getToken() -> String { VoqalFlutterPlugin.currentToken() }
    func getMetaData() -> String? { VoqalFlutterPlugin.currentMetadata() }
    func getViewController() -> UIViewController {
      UIApplication.shared.connectedScenes
        .compactMap { ($0 as? UIWindowScene)?.keyWindow?.rootViewController }
        .first?.voqalTopmost ?? UIViewController()
    }
    func voqalButton(didUploadRecording result: String) {}
    func voqalButton(didFailWith error: Error) {
      #if DEBUG
        NSLog("[VoqalFlutter] error: \(error)")
      #endif
    }
  }
}

private extension UIViewController {
  /// The controller actually on screen, so the sheet presents over Flutter's UI.
  var voqalTopmost: UIViewController {
    presentedViewController?.voqalTopmost ?? self
  }
}
